sound/soc/sof/pm.c

```diff
@@ -129,10 +130,32 @@ static int sof_restore_pipelines(struct snd_sof_dev *sdev)
 	list_for_each_entry(scontrol, &sdev->kcontrol_list, list) {
 
 		/* notify DSP of kcontrol values */
-		ret = snd_sof_ipc_set_comp_data(sdev->ipc, scontrol,
-						SOF_IPC_COMP_SET_VALUE,
-						SOF_CTRL_TYPE_VALUE_CHAN_SET,
-						scontrol->cmd);
+		switch (scontrol->cmd) {
+		case SOF_CTRL_CMD_VOLUME:
+			ipc_cmd = SOF_IPC_COMP_SET_VALUE;
+			ctrl_type = SOF_CTRL_TYPE_VALUE_CHAN_SET;
+			ret = snd_sof_ipc_set_comp_data(sdev->ipc, scontrol,
+							ipc_cmd, ctrl_type,
+							scontrol->cmd);
+			break;
+		case SOF_CTRL_CMD_BINARY:
+
+			/* Check if control data contains valid data.
+			 * SOF_ABI_MAGIC will not match if there is no data.
+			 */
+			ipc_cmd = SOF_IPC_COMP_SET_DATA;
+			ctrl_type = SOF_CTRL_TYPE_DATA_SET;
+			if (scontrol->control_data->data->magic == SOF_ABI_MAGIC)
```

Comment:
This line is apparently too long. The automatic checks complain about this, sorry.